tango/common/from_params.py

```diff
@@ -141,7 +141,21 @@ def infer_method_params(
             # `from __future__ import annotation`, the annotation will be a str,
             # so we need to resolve it using `get_type_hints` from the typing module.
             # See https://www.python.org/dev/peps/pep-0563/ for more info.
-            parameters[param_name] = param.replace(annotation=get_type_hints(method)[param_name])
+            try:
+                parameters[param_name] = param.replace(
+                    annotation=get_type_hints(method)[param_name]
+                )
+            except TypeError as e:
+                if "'type' object is not subscriptable" in str(e):
+                    new_e = TypeError(
+                        f"Could not determine the type annotations for `{method.__qualname__}`. Did you use the "
+                        "lowercase `dict`, `list`, `set`, or `tuple` in the type annotations instead of the "
+                        "uppercase `Dict`, `List`, `Set`, `Tuple`?"
```

Comment:
If I understand correctly, this is catching the case where the annotation is `dict` instead of, say, `Dict[str, Any]` or whatever, right? But `dict[str, Any]` is still okay, no? If so, I think we should make this error message more clear. Something like: "Did you forget to set the type's arguments in brackets? E.g. dict[str, Any]"

Reply:
`dict[str, Any]` is OK in the sense that you can put anything there and Python will ignore it anyways. But as soon as you try to use it it fails.

Reply:
> Python will ignore it anyways

What do you mean? Can `FromParams` not access the `str` and `Any` from the annotations?

> But as soon as you try to use it it fails.

Try to use it how? And how does it fail?

Reply:
>> Python will ignore it anyways

> What do you mean? Can `FromParams` not access the `str` and `Any` from the annotations?

Python, by default, does not do anything with the annotations. `FromParams` does, and promptly fails when the annotation contains `dict`.

> > But as soon as you try to use it it fails.

> Try to use it how? And how does it fail?

When you try to use it in `FromParams`. In detail, the call that doesn't work is `typing.get_type_hints()`. I wanted to give you an example that you can paste into a Python console, but it turns out that the Python console attempts to parse the type hints. Who knew?

```
>>> from typing import *
>>> def f(d: dict[str, str]) -> None:
...     pass
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'type' object is not subscriptable
```

That is the exact error you get in `FromParams` when it calls `get_type_hints()`.

Reply:
I guess I can trigger the error by making the type hint a string:

```
>>> def f(d: 'dict[str, str]') -> None:
...     pass
...
>>> get_type_hints(f)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/miniconda3/envs/meta-learn-prompt/lib/python3.7/typing.py", line 1013, in get_type_hints
    value = _eval_type(value, globalns, localns)
  File "/opt/miniconda3/envs/meta-learn-prompt/lib/python3.7/typing.py", line 263, in _eval_type
    return t._evaluate(globalns, localns)
  File "/opt/miniconda3/envs/meta-learn-prompt/lib/python3.7/typing.py", line 467, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
TypeError: 'type' object is not subscriptable
```

Reply:
This is the line: https://github.com/allenai/tango/blob/main/tango/common/from_params.py#L144

Reply:
You're running into these errors because you're using Python 3.7. `d: dict[str, str]` is not valid syntax in Python 3.7. It's only valid as of Python 3.10 or if you do `from __future__ import annotations`.

Reply:
** valid as of 3.9

Reply:
We already have a test for this: https://github.com/allenai/tango/blob/5d602fbf83998020288a986e2ef1b356e1c113ea/tests/common/from_params_test.py#L1038-L1052

Reply:
Should I change the error message to say "Please use `from __future__ import annotations`"?

Reply:
That would be good. But only if the Python version is < 3.9

Reply:
No, I checked this, that's not the case. When you don't use `from __future__ import annotations`, your whole file will fail to import. None of our code ever runs.

When you do use `from __future__ import annotations`, the file will import, but `param.annotation` is a `str`, which needs parsing. That's why we're in this branch in the first place. `get_type_hints()` does that parsing, but regardless of `annotations`, it will fail if your annotation is `dict`.

In Python >= 3.10, we will never even go into this branch.

The comment right above that part of the code is actually quite accurate and informative. Someone very smart must have written that.

Reply:
Okay, so correct me if I'm wrong:

First of all, none of this is an issue with Python 3.9 or newer. With Python 3.7 or 3.8 and `from __future__ import annotations`, a type hint like `d: dict` is technically okay (but not very informative), while a type hint like `d: dict[str, str]` will fail with "TypeError: 'type' object is not subscriptable."

In light of that, this makes sense to me now. Before, I was assuming that a type hint like `d: dict[str, str]` would still be okay when `__annotations__` was imported. That's not the case.

I think we can still make the error message a little more helpful. For one, it's not just `dict`, `list`, `set`, and `tuple` that this is an issue for, but really any type. Though, most likely, it will probably come from one of the many standard collections listed [here](https://www.python.org/dev/peps/pep-0585/).

See my other comment for a suggestion of how to improve it.